nh_eobs_mobile/controllers/main.py

```diff
@@ -874,10 +874,12 @@ def get_tasks(self, *args, **kw):
     @staticmethod
     def get_user_favourites(uid):
 
-        def location_not_false_or_already_exists():
-            if location and (location not in [x['location'] for x in favourites] or f.is_default):
-                return True
-            return False
+        def _remove_duplicates():
+            cleaned = [fav for fav in favourites if fav['default'] == 'true']
```

Comment:
You are entering twice in favourites loop, if you cant do in a single line both conditions i would suggest to do in the standard for loop.

Reply:
Please suggest an alternative method. Originally I had an issue of it not caring if the duplicate was 'default = true' or not. This was the shortest way I could come up with to make sure the defaults are captured first, and then add any remaining.

Reply:
Sample data:
  [{'location': 'Ward A', 'default': 'false'}, {'location': 'Ward B', 'default': 'ture'}, {'location': 'Ward A', 'default': 'true'}]

Reply:
I would consider :
cleaned=[]
for fav in favourites.sorted(key=default, reverse=True):
     if fav['location'] not in [loc['location'] for loc in cleaned]:
         cleaned.append(fav)

Reply:
Another possibility:

```
def _remove_duplicates(favourites):
   def sortkey_defaults_first(favourite):
        return 1 if favourite['default'] else 2
 
    cleaned = []
    locations_previously_found = set()
    for favourite in sorted(favourites, key=sortkey_defaults_first):
        location = favourite['location']
        if location not in locations_previously_found:
            locations_previously_found.add(location)
            cleaned.append(fav)
        return cleaned
```

This makes explicit from the start that we are deduplicating on location, and avoids nested comprehension execution (though optimisation wasn't the main intention it does make the worst case n*log(n) rather than n^2).

I also think naming the key like that makes it a bit less of a mental leap than having to remember that bools sort False first and work back from there when trying to find intent.

Reply:
In fact https://toolz.readthedocs.io/en/latest/api.html#toolz.itertoolz.unique can replace the loop and the set.
The key for that (not the sort) would be `itemgetter('location')`

```
from toolz.itertoolz import unique as unique_everseen
favourites_preferring_defaults=sorted(favourites, key=sortkey_defaults_first)
return list(unique_everseen(favourites_preferring_defaults, key=itemgetter('location')))
```